# General Purpose Parameters

# debug flags
verbose: false                      # Print additional debug and timing information.
visualize: true                     # Whether to publish information for visualization in RViz.

# frames
world_frame: world                  # The base frame used when looking up tf transforms. 
                                    # This is also the frame that most outputs are given in.
frame_id: world                     # Global fixed frame id.
local_frame_id: world               # Frame name for the quadrotor base.

# robot constraints
robot_radius: 0.45                  # Robot sphere radius, to be used for planning,  in metres.
traversability_radius: 0.45         # Radius in metres, of the space around the robot to be considered as occupied.
                                    # Recommended value: equal to robot_radius     

# esdf server params
publish_tsdf_map: false             # Whether to publish the complete TSDF map periodically over ROS topics.
publish_esdf_map: false             # Whether to publish the complete ESDF map periodically over ROS topics.
publish_pointclouds: false          # If true the tsdf and esdf (if generated) is published as a pointcloud when the mesh is updated 
publish_slices: false               # Whether to publish TSDF and ESDF pointcloud slices at the height specified by $slice_level
publish_traversable: false          # Whether to display a traversability pointcloud from the ESDF server.

update_mesh_every_n_sec: 0.0        # Rate at which the mesh topic will be published to, a value of 0 disables. 
                                    # This will not trigger any other mesh operations, such as generating a ply file.
update_esdf_every_n_sec: 0.25       # How often the ESDF map should be updated
publish_map_every_n_sec: 0.25       # Rate at which the TSDF and ESDF maps are published, if they are published.

# tsdf integrator params
method: fast                        # TSDF integrator method: can be 'simple', 'merged', or 'fast'
tsdf_voxel_size: 0.2                # Size of the tsdf voxels, in meters
tsdf_voxels_per_side: 16            # TSDF voxels per side of an allocated block. Must be a power of 2
esdf_max_distance_m: 2.0            # The maximum distance that the esdf will be calculated out to.
max_ray_length_m: 10.0              # The maximum range out to which a ray will be cast
voxel_carving_enabled: true         # If true, the entire length of a ray is integrated, 
                                    # if false only the region inside the truncation distance is used.
color_mode: height                  # The method that will be used for coloring the mesh. 
                                    # Options are "color", "height", "normals", "lambert" and "gray".

# esdf integrator params
use_tf_transforms: true             # If true, the ROS TF tree will be used to get the pose of the sensor 
                                    # relative to the world ($sensor_frame and $world_frame will be used). 
                                    # If false the pose must be given via the transform topic.
min_time_between_msgs_sec: 0.1      # Minimum time to wait after integrating a message before accepting a new one.
clear_sphere_for_planning: true     # The size of the queue used to subscribe to pointclouds.
occupied_sphere_radius: 4.0         # Radius of the outer sphere where unknown is set to occupied, in meters.
clear_sphere_radius: 0.1            # Radius of the inner sphere where unknown is set to free, in meters.
                                    # Recommended to set slightly more than the robot radius
slice_level: 1.0                    # The height at which generated tsdf and esdf slices will be made.
esdf_add_occupied_crust: true       # Whether to add an outside layer of occupied voxels. 
                                    # Basically just sets all unknown voxels in the allocated blocks to occupied.
esdf_min_diff_m: 0.0                # For cheaper but less accurate map updates: the minimum difference in
                                    # a voxel distance, before the change is propagated.

# global planner params
generate_by_layer_neighbors: true   # Whether to generate vertices/edges by number of basis points (false, default)
                                    # or number of neighbors on the discretized medial axis (true).
min_separation_angle: 0.78          # Minimum separation angle between two edges of a sparse graph node, in radians
                                    # If using full euclidean: 0.78 (45 degrees), If using quasi-Euclidean: 1.57 (90 degrees)

# local planner params
replan_dt: 0.1                      # Time to wait for before replanning, in seconds.
command_publishing_dt: 0.25         # Time between each command published by the local planner, in seconds.
replan_lookahead_sec: 1.0           # No. of seconds ahead to consider if the need for replanning arises.
mpc_prediction_horizon: 30          # No. of timesteps to plan ahead for
planning_horizon_m: 20.0            # Planning horizon, in metres.
v_max: 2.0                          # Max allowed velocity, in metres per second.
a_max: 2.0                          # Max allowed accelration, in metres per seconds squared.
autostart: true                     # Whether to auto-start publishing any new path or wait for start service call.
use_shotgun: false                  # Find the intermediate point by shotgun firing if the planning horizon is occupied
plan_to_start: true                 # Whether to start planning at the current odometry.
avoid_collisions: true              # Whether to reject paths that collide with the environment.
goal_selector_strategy: local       # What to do in case planning fails. Options: 'local', 'random', 'none'